File: programmers/Python/P159994_카드_뭉치.py
"""
Queue 자료구조 사용; O(N + M)
- card1과 card2는 무조건 앞부터 사용해야 함.
- 순서를 뒤바꿀 수 없음.
"""
from collections import deque

def solution(cards1, cards2, goal):
    q_cards1 = deque(cards1)
    q_cards2 = deque(cards2)
    q_goal = deque(goal)
    
    while q_goal:
        cur = q_goal.popleft()
        
        if not q_cards1 and not q_cards2:
            return "No"
        
        if q_cards1 and cur == q_cards1[0]:
            q_cards1.popleft()
        elif q_cards2 and cur == q_cards2[0]:
            q_cards2.popleft()
        else:
            return "No"
    
    return "Yes"

File: programmers/Java/P159994_카드_뭉치.java
/**
Queue 자료구조 사용; O(N + M)
- card1과 card2는 무조건 앞부터 사용해야 함.
- 순서를 뒤바꿀 수 없음.
*/
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    public String solution(String[] cards1, String[] cards2, String[] goal) {
        Queue<String> qCards1 = new LinkedList<>();
        Queue<String> qCards2 = new LinkedList<>();
        Queue<String> qGoal = new LinkedList<>();
        
        // 카드 리스트를 큐로 변환
        for (String card : cards1)
            qCards1.add(card);

        for (String card : cards2)
            qCards2.add(card);
        
        for (String word : goal)
            qGoal.add(word);
        
        while (!qGoal.isEmpty()) {
            String cur = qGoal.poll();
            
            if (!qCards1.isEmpty() && cur.equals(qCards1.peek())) {
                qCards1.poll();
            } else if (!qCards2.isEmpty() && cur.equals(qCards2.peek())) {
                qCards2.poll();
            } else {
                return "No";
            }
        }
        return "Yes";
    }
}
